karapace/protobuf/io.py

```diff
@@ -0,0 +1,206 @@
+    directory = cfg["protobuf_runtime_directory"]
+    proto_name = calculate_class_name(str(schema))
+    proto_path = f"{directory}/{proto_name}.proto"
+    class_path = f"{directory}/{proto_name}_pb2.py"
+    if not os.path.isfile(proto_path):
+        with open(f"{directory}/{proto_name}.proto", "w") as proto_text:
+            proto_text.write(str(schema))
+
+    if not os.path.isfile(class_path):
+        subprocess.run([
+            "protoc",
+            "--python_out=./",
+            proto_path,
+        ], check=True)
+
+    spec = importlib.util.spec_from_file_location(f"{proto_name}_pb2", class_path)
+    tmp_module = importlib.util.module_from_spec(spec)
+    spec.loader.exec_module(tmp_module)
+    class_to_call = getattr(tmp_module, class_name)
+    return class_to_call()
+
+
+def read_data(writer_schema: ProtobufSchema, reader_schema: ProtobufSchema, bio: BytesIO):
+    # TODO (serge): check and polish it
+    if not match_schemas(writer_schema, reader_schema):
+        fail_msg = 'Schemas do not match.'
+        raise ProtobufSchemaResolutionException(fail_msg, writer_schema, reader_schema)
+
+    indexes = read_indexes(bio)
+    name = find_message_name(writer_schema, indexes)
+
+    class_instance = get_protobuf_class_instance(writer_schema, name, config.DEFAULTS)
+    class_instance.ParseFromString(bio.read())
+
+    return class_instance
+
+
+class ProtobufDatumReader:
+    """Deserialize Protobuf-encoded data into a Python data structure."""
+
+    def __init__(self, writer_schema=None, reader_schema=None):
+        """ As defined in the Protobuf specification, we call the schema encoded
+        in the data the "writer's schema", and the schema expected by the
+        reader the "reader's schema".
+        """
+        self._writer_schema = writer_schema
+        self._reader_schema = reader_schema
+
+    def read(self, bio: BytesIO):
+        if self._reader_schema is None:
+            self._reader_schema = self._writer_schema
+        return protobuf_to_dict(read_data(self._writer_schema, self._reader_schema, bio), True)
+
+
+def write_varint(bio: BytesIO, value: int) -> int:
+    if value < 0:
+        raise ValueError(f"value must not be negative, got {value}")
+
+    if value == 0:
+        bio.write(ZERO_BYTE)
+        return 1
+
+    written_bytes = 0
+    while value > 0:
+        to_write = value & 0x7f
+        value = value >> 7
+
+        if value > 0:
+            to_write |= 0x80
+
+        bio.write(bytearray(to_write)[0])
+        written_bytes += 1
+
+    return written_bytes
+
+
+def write_indexes(bio: BytesIO, indexes: List[int]) -> None:
+    for i in indexes:
+        write_varint(bio, i)
+
+
+class ProtobufDatumWriter:
+    """ProtobufDatumWriter for generic python objects."""
+
+    def __init__(self, writer_schema=None):
+        self._writer_schema = writer_schema
+        a: ProtobufSchema = writer_schema
```

Comment:
This is unnecessary, because the constructor argument shouldn't be optional.

Reply:
optional constructor arguments come from template See. avro/io.py